lib/Auth/Process/LogUser.php

```diff
@@ -0,0 +1,224 @@
+     * Apply filter to copy attributes.
+     *
+     * @param array &$state  The current state array
+     */
+    public function process(&$state) {
+        if (! $this->configsAreValid()) {
+            return;
+        }
+
+        assert('is_array($state)');
+
+        $idp = $this->getIdp($state);
+
+        // Get the SP's entity id
+        $spEntityId = "SP entity ID not available";
+        if (! empty($state['saml:sp:State']['SPMetadata']['entityid'])) {
+            $spEntityId = $state['saml:sp:State']['SPMetadata']['entityid'];
+        }
+
+        // Get the current datetime
+        $datetime = date("Y-m-d H:i:s");
+
+        $sdk = new Aws\Sdk([
+            'endpoint'   => $this->dynamoEndpoint,
+            'region'   => $this->dynamoRegion,
+            'version'  => 'latest'
+        ]);
+
+        $dynamodb = $sdk->createDynamoDb();
+        $marshaler = new Marshaler();
+
+        $userAttributes = $this->getUserAttributes($state);
+
+        $logJson = '
+            {
+                "ID": "' . uniqid() . '",
+                "IDP": "' . $idp . '",
+                "SP": "' . $spEntityId . '",' .
+                $userAttributes .
+                '"Time": "' . $datetime . '"
+            }';
+
+        $item = $marshaler->marshalJson($logJson);
+
+        $params = [
+            'TableName' => $this->dynamoLogTable,
+            'Item' => $item,
+        ];
+
+        try {
+            $result = $dynamodb->putItem($params);
+        } catch (DynamoDbException $e) {
+            echo "Unable to add item:\n";
+            echo $e->getMessage() . "\n";
+        }
+    }
+
+    private function configsAreValid() {
+        $msg = ' config value not provided to LogUser.';
+        if (empty($this->dynamoEndpoint)) {
+            SimpleSAML\Logger::error(self::DYNAMO_ENDPOINT_KEY . $msg);
+            return false;
+        }
+
+        if (empty($this->dynamoRegion)) {
+            SimpleSAML\Logger::error(self::DYNAMO_REGION_KEY . $msg);
+            return false;
+        }
+
+        if (empty($this->dynamoLogTable)) {
+            SimpleSAML\Logger::error(self::DYNAMO_LOG_TABLE_KEY . $msg);
+            return false;
+        }
+
+        return true;
+    }
+
+    private function getIdp(&$state) {
+        if (empty($state[self::IDP_KEY])) {
+            return 'No IDP available';
+        }
+
+        $samlIDP = $state[self::IDP_KEY];
+
+        // Get the potential IDPs from idp remote metadata
+        $metadataPath = __DIR__ . '/../../../../../metadata';
+
+        // If a unit test sends a different metadataPath, use it
+        if (isset($state['metadataPath'])) {
+            $metadataPath = $state['metadataPath'];
+        }
+        $idpEntries = \Sil\SspUtils\Metadata::getIdpMetadataEntries($metadataPath);
+
+        // Get the IDPNamespace or else just use the IDP's entity ID
+        $idpEntry = $idpEntries[$samlIDP];
+
+        // If the IDPNamespace entry is close to being alphanumeric, use it
```

Comment:
Why does it matter whether the IDP Namespace is close to being alphanumeric?

Reply:
Fixed